### ISSUE TITLE
Fetch all should fetch tags as well

### ISSUE THREAD

**Issue by drauch** [gitextensions/gitextensions#7094]
## Feature description

1) It should be configurable whether "Fetch all" fetches tags as well. Currently it looks like, in order to fetch the tags too, you have to use the "Open pull dialog..." option.

2) It should be configurable (both, for the feature described in (1) and in "Open pull dialog...") whether fetching the tags should use the "--force" option to overwrite locally existing tags.

## Environment

- Git Extensions 3.1.1.6049
- Build 2f872105b4eb42421d9fdafd68e90a9e27d20d96
- Git 2.22.0.windows.1
- Microsoft Windows NT 10.0.17134.0
- .NET Framework 4.7.3416.0
- DPI 96dpi (no scaling)

Best regards,
D.R.

**Comment by RussKie** [comment]
> 1. "Fetch all" fetches tags as well.

"Fetch all" fetches all remote refs including tags.

**Comment by drauch** [comment]
It does not for me.

1) If I do fetch all -> tags are not fetched, no warning about scrambling tags.
2) If I do "Open pull dialog..." -> "Fetch all tags" -> I get the following error message:
{code}
"C:\Program Files\Git\bin\git.exe" fetch --progress "origin" --tags
From https://git.int.rubicon-it.com/scm/rnd_css/signpath.application
 ! [rejected]            HealthCheck-sprint-release-latest -> HealthCheck-sprint-release-latest  (would clobber existing tag)
 ! [rejected]            HealthCheck-sprint-release-latest-fqa -> HealthCheck-sprint-release-latest-fqa  (would clobber existing tag)
{code}
Maybe these two messages are swallowed in (1) ? Still weird.

But to overcome this problem, I'm really looking forward to (2).

**Comment by tschlue** [comment]
The same for the 'fetch'-dialog. The quicknote says 'Fetch branches and tags' but the command is just 'git fetch --progress "origin"'.

**Comment by JasperCiti** [comment]
Confirmed. I ran into the same issue. Git Extensions only does `git fetch --progress "--all"` but when I manuallly ran `git fetch --all --tags` I got my tags.

**Comment by ooxxTaiwan** [comment]
so, any update for this?
![image](https://user-images.githubusercontent.com/3377713/100995186-6ae7b700-3592-11eb-8d3e-77e5f4e2bb70.png)

**Comment by RussKie** [comment]
This is working in 3.4.3:
![image](https://user-images.githubusercontent.com/4403806/101001888-cf673e00-35b3-11eb-9992-f2194e6094dd.png)

You can also fetch tags from the dialog:
![image](https://user-images.githubusercontent.com/4403806/101001215-f2ddb900-35b2-11eb-9b8f-b45ca5ca920b.png)

**Comment by ivangrek** [comment]
Please see config. May be you disabled this feature.

https://git-scm.com/docs/git-fetch#Documentation/git-fetch.txt---tags

**Comment by ooxxTaiwan** [comment]
> Confirmed. I ran into the same issue. Git Extensions only does `git fetch --progress "--all"` but when I manuallly ran `git fetch --all --tags` I got my tags.

@RussKie 
I got the same problem. any idea? 
thanks.

**Comment by ooxxTaiwan** [comment]
> Please see config. May be you disabled this feature.
> 
> https://git-scm.com/docs/git-fetch#Documentation/git-fetch.txt---tags

@ivangrek 
How to check "disabled this feature" or not?
Thanks.

**Comment by drauch** [comment]
> This is working in 3.4.3:

I want it to _always_ fetch all tags _in addition_ to fetching the rest on each fetch. How to configure that? When I select "Fetch all tags" it does not remember the setting. Next time I click "Fetch all" it does not fetch them ...

**Comment by RussKie** [comment]
In the 8+ years of using GE "fetch all" was always fetching tags for me
(including work against GitHub, BitBucket, both public and enterprise
instances, and local).

So I am really puzzled why it is not your experience.

**Comment by ivangrek** [comment]
git config --list

remote.<name>.tagOpt

`By default, tags that point at objects that are downloaded from the remote repository are fetched and stored locally. This option disables this automatic tag following. The default behavior for a remote may be specified with the remote.<name>.tagOpt setting. See git-config[1].`

https://git-scm.com/docs/git-fetch#Documentation/git-fetch.txt---no-tags

**Comment by ivangrek** [comment]
Also see `remote.<repository>.fetch`  I think refspec may change behavior too.

**Comment by drauch** [comment]
`git config --list` is not outputting `remote.<anything>.tagOpt` and `remote.origin.fetch` is set to `+refs/heads/*:refs/remotes/origin/*` - is this a problem? What do you have @RussKie?

**Comment by RussKie** [comment]
Here's a snip:
```
core.autocrlf=true
core.fscache=true
core.symlinks=false
pull.rebase=false
core.safecrlf=true
core.fscache=true
core.autocrlf=true
pull.rebase=false
fetch.prune=false
fetch.writecommitgraph=true
rebase.autostash=false
core.repositoryformatversion=0
core.filemode=false
core.bare=false
core.logallrefupdates=true
core.symlinks=false
core.ignorecase=true
core.autocrlf=true
core.safecrlf=false
submodule.active=.
remote.remote.url=https://github.com/gitextensions/gitextensions.git
remote.remote.fetch=+refs/heads/*:refs/remotes/remote/*
remote.origin.url=https://github.com/RussKie/gitextensions.git
remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
```

**Comment by drauch** [comment]
Hmm, that doesn't look different at my machine...if you look at the command log after pressing fetch all in GitExtensions, is your entry different to my `"C:\Program Files\Git\bin\git.exe" fetch --progress "--all"`?

**Comment by mstv** [comment]
No, it is exactly the same for me. And it fetches the release tags. Weird, why it does not work for you.
Though your git is a little old.
Have you tried to clone it again to a new folder?
You could also try it with your `%USERPROFILE%\.gitconfig` renamed temporarily.

**Comment by drauch** [comment]
* I've since updated my Git to 2.28.0.windows.1, so it's most probably not due to my older version of Git.
* I'm using GitExtensions on Windows
* Cloning the repo into a new folder didn't help (but as it's not doing it in any repo, I don't think that's the problem. I'm regularly checking out new repositories)
* Renaming my .gitconfig temporarily didn't help either

**Comment by vbjay** [comment]
Reply with a full listing of all git configs


git config --system -l


git config --global -l


git config -l


Redact any sensitive config data. 


Then clone one of your repos and make sure it has tags. Delete a tag (ONLY LOCAL). Then clear the ge command log and then do a fetch all. Send the log you get from ge. 


If you moved a tag in remote repo and your local repo is pointing to a different commit, then you may have trouble. Tags shouldn't be bounced around. 

Also are you running in wine or other emulation software or straight windows?

**Comment by drauch** [comment]
System:
```
http.sslcainfo=C:/Program Files/Git/mingw64/ssl/certs/ca-bundle.crt
http.sslbackend=openssl
diff.astextplain.textconv=astextplain
filter.lfs.clean=git-lfs clean -- %f
filter.lfs.smudge=git-lfs smudge -- %f
filter.lfs.process=git-lfs filter-process
filter.lfs.required=true
credential.helper=manager
core.editor="C:\\Program Files (x86)\\Notepad++\\notepad++.exe" -multiInst -notabbar -nosession -noPlugin
core.autocrlf=true
core.fscache=true
core.symlinks=true
pull.rebase=false
````
Global:
```
core.editor="C:/Program Files (x86)/notepad++/notepad++.exe" -multiInst -nosession
core.preloadindex=true
core.fscache=true
diff.compactionheuristic=true
diff.guitool=beyondcompare4
diff.renamelimit=100000
difftool.beyondcompare4.path=C:/Program Files/Beyond Compare 4/bcomp.exe
difftool.beyondcompare4.cmd="C:/Program Files/Beyond Compare 4/bcomp.exe" "$LOCAL" "$REMOTE"
http.sslverify=true
merge.tool=BeyondCompare4
merge.renamelimit=100000
merge.guitool=BeyondCompare4
mergetool.BeyondCompare4.path=C:/Program Files/Beyond Compare 4/bcomp.exe
mergetool.BeyondCompare4.cmd="C:/Program Files/Beyond Compare 4/bcomp.exe" "$LOCAL" "$REMOTE" "$BASE" "$MERGED"
user.useconfigonly=true
pull.rebase=false
fetch.prune=false
rebase.autostash=false
```
Local:
```
http.sslcainfo=C:/Program Files/Git/mingw64/ssl/certs/ca-bundle.crt
http.sslbackend=openssl
diff.astextplain.textconv=astextplain
filter.lfs.clean=git-lfs clean -- %f
filter.lfs.smudge=git-lfs smudge -- %f
filter.lfs.process=git-lfs filter-process
filter.lfs.required=true
credential.helper=manager
core.editor="C:\\Program Files (x86)\\Notepad++\\notepad++.exe" -multiInst -notabbar -nosession -noPlugin
core.autocrlf=true
core.fscache=true
core.symlinks=true
pull.rebase=false
core.editor="C:/Program Files (x86)/notepad++/notepad++.exe" -multiInst -nosession
core.preloadindex=true
core.fscache=true
diff.compactionheuristic=true
diff.guitool=beyondcompare4
diff.renamelimit=100000
difftool.beyondcompare4.path=C:/Program Files/Beyond Compare 4/bcomp.exe
difftool.beyondcompare4.cmd="C:/Program Files/Beyond Compare 4/bcomp.exe" "$LOCAL" "$REMOTE"
http.sslverify=true
merge.tool=BeyondCompare4
merge.renamelimit=100000
merge.guitool=BeyondCompare4
mergetool.BeyondCompare4.path=C:/Program Files/Beyond Compare 4/bcomp.exe
mergetool.BeyondCompare4.cmd="C:/Program Files/Beyond Compare 4/bcomp.exe" "$LOCAL" "$REMOTE" "$BASE" "$MERGED"
user.useconfigonly=true
pull.rebase=false
fetch.prune=false
rebase.autostash=false
core.repositoryformatversion=0
core.filemode=false
core.bare=false
core.logallrefupdates=true
core.ignorecase=true
user.name=<removed>
user.email=<removed>
submodule.active=.
remote.origin.url=<removed>
remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
branch.master.remote=origin
branch.master.merge=refs/heads/master
submodule.Shared.url=<removed>
gui.wmstate=normal
gui.geometry=893x435+78+78 175 196
<lots_of_branches>
```

Command log after fetch all:
```
2020-12-08T10:18:45.8729143+01:00	49	55808	UI	0	git	ls-files -z --unmerged	C:\Path\To\Repo\
2020-12-08T10:18:45.9599112+01:00	46	36452	UI	0	git	remote	C:\Path\To\Repo\
2020-12-08T10:18:46.1114467+01:00	46	48688	UI	0	git	rev-parse --git-path "shallow"	C:\Path\To\Repo\
2020-12-08T10:18:46.1943079+01:00	3225	54764	UI		git	fetch --progress "--all"	C:\Path\To\Repo\
2020-12-08T10:18:49.4369438+01:00	47	48280	UI	0	git	ls-files -z --unmerged	C:\Path\To\Repo\
2020-12-08T10:18:49.4858225+01:00	124	37796		0	git	show-ref --tags	C:\Path\To\Repo\
2020-12-08T10:18:49.4858225+01:00	124	40884		0	git	--no-optional-locks for-each-ref --sort=-committerdate refs/heads/ --format="%(objectname) %(refname)"	C:\Path\To\Repo\
2020-12-08T10:18:49.4858225+01:00	112	27260		0	git	--no-optional-locks show-ref --dereference	C:\Path\To\Repo\
2020-12-08T10:18:49.5096525+01:00	108	42800	UI	0	git	rev-parse HEAD	C:\Path\To\Repo\
2020-12-08T10:18:49.6108730+01:00	58	54076		0	git	remote -v	C:\Path\To\Repo\
2020-12-08T10:18:49.6108730+01:00	508	35276		0	git	for-each-ref --format="%(push:track)::%(upstream:track)::%(refname:short)" refs/heads/	C:\Path\To\Repo\
2020-12-08T10:18:49.6420207+01:00	62	16396	UI	0	git	show-ref --dereference	C:\Path\To\Repo\
2020-12-08T10:18:49.6795974+01:00	71	43432		0	git	show-ref --dereference	C:\Path\To\Repo\
2020-12-08T10:18:49.7236301+01:00	450	24188			git	-c log.showSignature=false log -z --pretty=format:"%H%T%P%n%at%n%ct%n%e%n%aN%n%aE%n%cN%n%cE%n%s%n%n%b1DEA7CC4-FB39-450A-8DDF-762FCEA28B05" --all --boundary --not --glob=notes --not --max-count=100000 --	C:\Path\To\Repo\
2020-12-08T10:18:49.7295966+01:00	121	36492		0	git	for-each-ref --sort=-committerdate refs/heads/ --format="%(objectname) %(refname)"	C:\Path\To\Repo\
2020-12-08T10:18:49.7295966+01:00	165	45936	UI	0	git	branch --format=%(refname) -a --merged HEAD	C:\Path\To\Repo\
2020-12-08T10:18:49.7465958+01:00	94	31504	UI	0	git	ls-files -z --unmerged	C:\Path\To\Repo\
2020-12-08T10:18:49.7605955+01:00	104	49716		0	git	remote	C:\Path\To\Repo\
2020-12-08T10:18:49.8745963+01:00	61	40932	UI	0	git	for-each-ref --format="%(push:track)::%(upstream:track)::%(refname:short)" refs/heads/<current-feature-branch-i'm-on>	C:\Path\To\Repo\
2020-12-08T10:18:50.5521157+01:00	112	48924	UI	0	git	--no-optional-locks -c diff.ignoreSubmodules=none status --porcelain=2 -z --untracked-files	C:\Path\To\Repo\
2020-12-08T10:18:50.5641519+01:00	107	23284	UI	0	git	branch --format=%(refname) -a --merged 281c08b1027d03b6f7f6273aff303f012a351704	C:\Path\To\Repo\
2020-12-08T10:18:50.5641519+01:00	83	52588		0	git	branch --contains 281c08b1027d03b6f7f6273aff303f012a351704	C:\Path\To\Repo\
2020-12-08T10:18:50.6575610+01:00	54	37812		0	git	tag --contains 281c08b1027d03b6f7f6273aff303f012a351704	C:\Path\To\Repo\
2020-12-08T10:18:50.7521904+01:00	56	41384		0	git	describe --tags --first-parent --abbrev=40 281c08b1027d03b6f7f6273aff303f012a351704	C:\Path\To\Repo\
```

> Also are you running in wine or other emulation software or straight windows?
Nope, straight Windows 10 1909

Best regards,
D.R.

**Comment by RussKie** [comment]
Another machine/another repo:
![image](https://user-images.githubusercontent.com/4403806/101562063-b37ff400-3a1a-11eb-857a-6e201b1e6196.png)

@drauch are you able to try it on another machine perhaps?

**Comment by mstv** [comment]
Does the output of `git show-ref --dereference` end with something like the following as for the GE repo?
Does it contain special characters perhaps?
```
e8a8f903f2dfe62dea9fbc0609a3aeab916edde2 refs/tags/v3.4.2
082cfd97d09c47346638b1873f5c8cf6ed139382 refs/tags/v3.4.2^{}
3cbdf1e3817760f27aec31fb9ca1df2eddc26698 refs/tags/v3.4.3
d4b0f48bbf3e39a71f5d7ff5231be5678d4aa0f1 refs/tags/v3.4.3^{}
```

Though I read the issue description and the first comments again. This issue seems to be about conflicting tags, isn't it, @drauch?
And about swallowed git warnings.
But not about cloning a repo into a new folder.

**Comment by vbjay** [comment]
That would be my guess. I'll bet they moved tags and expect the tags to
move.

**Comment by drauch** [comment]
> I'll bet they moved tags and expect the tags to move.

Yes, of course, that's what I posted already in my initial post (see (2) of my original post). I want to force the tags from the remote repository on my repo on each fetch all.

**Comment by vbjay** [comment]
Open the pull dialog and look at the ui. You can cause this ti happen.

Tags are not meant to be moved around. They are designed to be markers. If
remote tag points to c2 and local tag points to c1 fetch all should not by
default replace the tag.   Use the dialog. If you are bouncing tags then
look at using a branch instead.

**Comment by vbjay** [comment]
https://git-scm.com/docs/git-config#Documentation/git-config.txt-remoteltnamegttagOpt

**Comment by drauch** [comment]
> Tags are not meant to be moved around. They are designed to be markers. If
remote tag points to c2 and local tag points to c1 fetch all should not by
default replace the tag.   Use the dialog. If you are bouncing tags then
look at using a branch instead.

It should not be the "by-default default", that's ok, but I want to make it the default for fetching at our company or at least on my machine. Our company uses tags as markers for currently deployed versions and they change all the time.

**Comment by pmiossec** [comment]
>Our company uses tags as markers for currently deployed versions and they change all the time.

From what I can understand, you are using git in a way that is not intended.
So why do you want GitExtensions to do something that git doesn't do.
It is not in GE job to do some strange things.

Perhaps you'd better have to create a GE script run before fetch that delete your "moving" tags so that git is able to fetch them again...

**Comment by vbjay** [comment]
So use the config option I linked to.

**Comment by vbjay** [comment]
Also why do you think we have semantic versioning?
https://semver.org/

**Comment by mstv** [comment]
The following user script should work. `cmd` and `echo` can be omitted. Do not let it run in the background in order to refresh the RevisionGrid afterwards.

![grafik](https://user-images.githubusercontent.com/36601201/101959693-1aadcc00-3c06-11eb-9139-d3c60f6b0f64.png)

**Comment by vbjay** [comment]
No need for cmd c

Just mage command git.

**Comment by vbjay** [comment]
remote.<name>.tagOpt

Setting this value to --no-tags disables automatic tag following when
fetching from remote <name>. Setting it to --tags will fetch every tag from
remote <name>, even if they are not reachable from remote branch heads.
Passing these flags directly to git-fetch[1]
<https://git-scm.com/docs/git-fetch> can override this setting. See options
--tags and --no-tags of git-fetch[1] <https://git-scm.com/docs/git-fetch>

**Comment by mstv** [comment]
@gitextensions/git-extensions-team, the previous comment gives a valid rationale to simply add `--force` to the `git fetch all` command - IMO - rather than adding yet another option or a button for setting some configuration option.
Any objections?

**Comment by mstv** [comment]
Perhaps, rather to the `Fetch and prune all`.

**Comment by RussKie** [comment]
No objections from me. I think it will compliment https://github.com/gitextensions/gitextensions/issues/8188.
👍

**Comment by mstv** [comment]
TL;DR: The combination of `git config remote.<name>.tagOpt --tags` - set only for the remote which hosts the tags - and of `git fetch --progress --all --prune --force` works well. Thanks, @vbjay.

---

TODO:
- add `--force` to `--prune`
- optional: add checkbox to the Form `Remote Repositories Manager` in order to set `tagOpt`

---

In order to update the tags, both arguments `--tags` and `--force` must be added to `git fetch`.
BUT: `Fetch all` means - at least in GE - to fetch _all_ remotes. If the tags are not in sync on the remotes, the tags are shifted back and forth on every fetch:

```
$ "C:\Program Files\Git\bin\git.exe" fetch --progress "--all" --tags --force
Fetching upstream
From https://github.com/gitextensions/gitextensions
 t [tag update]          v3.2       -> v3.2
 t [tag update]          v3.5-RC1   -> v3.5-RC1
Fetching russkie
From https://github.com/russkie/gitextensions
 t [tag update]          v3.2       -> v3.2
 t [tag update]          v3.5-RC1   -> v3.5-RC1
```

In combination with `--prune`, `--tags` is even worse if the tags have not been pushed to one of the forks.

That's why `--tags` should not be used for any variant of `Fetch all`.
It could be added to the plain `Fetch` and/or to a new `Fetch and prune`.
